File: backend/app/models/project_tracking.py
"""
项目跟踪相关数据库模型
"""
from sqlalchemy import Column, Integer, String, Date, DateTime, Boolean, Numeric, Text
from sqlalchemy.sql import func
from app.database import Base

class ProjectTracking(Base):
    """项目跟踪主表"""
    __tablename__ = "project_trackings"
    
    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(String(50), nullable=False, index=True)
    project_name = Column(String(200), nullable=False)
    
    # 跟踪状态
    overall_progress = Column(Numeric(5, 2), default=0.00)
    tracking_status = Column(String(20), default='进行中', index=True)
    last_update_time = Column(DateTime, default=func.now())
    
    # 统计信息
    total_tasks = Column(Integer, default=0)
    completed_tasks = Column(Integer, default=0)
    total_reports = Column(Integer, default=0)
    cde_evaluations = Column(Integer, default=0)
    
    # 风险等级
    risk_level = Column(String(10), default='低', index=True)
    priority_level = Column(String(10), default='中', index=True)
    
    # 时间戳
    create_time = Column(DateTime, default=func.now())
    update_time = Column(DateTime, default=func.now(), onupdate=func.now())
    is_deleted = Column(Boolean, default=False, index=True)


class TaskTracking(Base):
    """任务跟踪表"""
    __tablename__ = "task_trackings"
    
    id = Column(Integer, primary_key=True, index=True)
    tracking_id = Column(Integer, nullable=False, index=True)
    
    # 任务基本信息
    task_id = Column(String(50), nullable=False, index=True)
    task_name = Column(String(200), nullable=False)
    assignee = Column(String(100))
    assignee_id = Column(String(50))
    
    # 计划信息
    planned_start = Column(Date)
    planned_end = Column(Date)
    
    # 实际跟踪信息
    actual_start = Column(Date)
    actual_end = Column(Date)
    current_progress = Column(Numeric(5, 2), default=0.00)
    delay_days = Column(Integer, default=0)
    
    # 关联信息
    related_reports_count = Column(Integer, default=0)
    cde_evaluation_count = Column(Integer, default=0)
    
    # 状态跟踪
    status = Column(String(20), default='未开始', index=True)
    priority_level = Column(String(10), default='中')
    risk_level = Column(String(10), default='低')
    
    # 时间戳
    create_time = Column(DateTime, default=func.now())
    update_time = Column(DateTime, default=func.now(), onupdate=func.now())
    is_deleted = Column(Boolean, default=False, index=True)


class TaskReportLink(Base):
    """任务日报关联表"""
    __tablename__ = "task_report_links"
    
    id = Column(Integer, primary_key=True, index=True)
    task_tracking_id = Column(Integer, nullable=False, index=True)
    
    # 关联的日报信息
    report_id = Column(Integer, nullable=False, index=True)
    work_item_id = Column(Integer, nullable=False, index=True)
    
    # 关联详情
    employee_id = Column(String(50), nullable=False, index=True)
    employee_name = Column(String(100), nullable=False)
    report_date = Column(Date, nullable=False, index=True)
    
    # 工作详情
    work_content = Column(Text, nullable=False)
    hours_spent = Column(Numeric(8, 2), default=0.00)
    progress_contribution = Column(Numeric(5, 2), default=0.00)
    
    # 时间戳
    create_time = Column(DateTime, default=func.now())
    update_time = Column(DateTime, default=func.now(), onupdate=func.now())
    is_deleted = Column(Boolean, default=False, index=True)


class ProjectReportSummary(Base):
    """项目日报汇总表"""
    __tablename__ = "project_report_summaries"
    
    id = Column(Integer, primary_key=True, index=True)
    tracking_id = Column(Integer, nullable=False, index=True)
    
    # 汇总期间
    summary_date = Column(Date, nullable=False, index=True)
    
    # 人员统计
    involved_employees = Column(Integer, default=0)
    total_work_hours = Column(Numeric(10, 2), default=0.00)
    
    # 任务进展
    tasks_progressed = Column(Integer, default=0)
    tasks_completed = Column(Integer, default=0)
    
    # CDE评价统计
    cde_evaluations_count = Column(Integer, default=0)
    cde_evaluations_detail = Column(Text)
    
    # 问题和风险
    issues_raised = Column(Integer, default=0)
    risks_identified = Column(Integer, default=0)
    
    # 汇总内容
    daily_summary = Column(Text)
    next_day_plan = Column(Text)
    
    # 时间戳
    create_time = Column(DateTime, default=func.now())
    update_time = Column(DateTime, default=func.now(), onupdate=func.now())
    is_deleted = Column(Boolean, default=False, index=True)


class TrackingNotification(Base):
    """跟踪通知表"""
    __tablename__ = "tracking_notifications"
    
    id = Column(Integer, primary_key=True, index=True)
    tracking_id = Column(Integer, nullable=False, index=True)
    
    # 通知类型
    notification_type = Column(String(20), nullable=False, index=True)
    priority_level = Column(String(10), default='中')
    
    # 通知内容
    title = Column(String(200), nullable=False)
    content = Column(Text, nullable=False)
    related_task_id = Column(String(50))
    related_report_id = Column(Integer)
    
    # 接收人信息
    recipient_id = Column(String(50), nullable=False, index=True)
    recipient_name = Column(String(100), nullable=False)
    recipient_role = Column(String(50))
    
    # 通知状态
    is_read = Column(Boolean, default=False, index=True)
    is_sent = Column(Boolean, default=False, index=True)
    sent_time = Column(DateTime)
    read_time = Column(DateTime)
    
    # 推送方式
    push_methods = Column(String(100))
    
    # 时间戳
    create_time = Column(DateTime, default=func.now())
    update_time = Column(DateTime, default=func.now(), onupdate=func.now())
    is_deleted = Column(Boolean, default=False, index=True)

File: frontend/src/api/projectTracking.ts
import axios from './axios'

// 项目跟踪API接口 - 基于优化后的项目跟踪表结构
export const projectTrackingApi = {
  // 获取项目跟踪列表 - 直接查询原始表
  getProjectTrackings(params?: {
    project_name?: string
    status?: string
    page?: number
    limit?: number
    sort_by?: string
    sort_order?: 'asc' | 'desc'
  }) {
    return axios.get('v1/project-tracking-list', { params })
  },

  // 获取项目跟踪详情 - 包含所有关联数据
  getProjectTrackingDetail(trackingId: number) {
    return axios.get(`v1/project-tracking/projects/${trackingId}/detail`)
  },

  // 获取项目甘特图数据
  getProjectGanttData(projectId: string) {
    return axios.get(`v1/project-tracking/${projectId}/gantt`)
  },

  // 获取项目任务跟踪列表 - 基于task_trackings表
  getTaskTrackings(trackingId: number) {
    return axios.get(`v1/project-tracking/projects/${trackingId}/tasks`)
  },

  // 获取CDE评价信息 - 基于优化后的task_cde_evaluations视图
  getCDEEvaluations(trackingId: number, params?: {
    start_date?: string
    end_date?: string
    min_score?: number
    max_score?: number
  }) {
    return axios.get(`v1/project-tracking/projects/${trackingId}/cde-evaluations`, { params })
  },

  // 获取进度趋势数据 - 用于图表展示
  getProgressTrend(trackingId: number, params?: {
    period?: 'week' | 'month' | 'quarter'
    start_date?: string
    end_date?: string
  }) {
    return axios.get(`v1/project-tracking/projects/${trackingId}/progress-trend`, { params })
  },

  // 获取风险分析数据
  getRiskAnalysis(trackingId: number) {
    return axios.get(`v1/project-tracking/projects/${trackingId}/risk-analysis`)
  },

  // 更新项目跟踪状态
  updateProjectTracking(trackingId: number, data: {
    tracking_status?: string
    overall_progress?: number
    risk_level?: string
    priority_level?: string
    notes?: string
  }) {
    return axios.put(`v1/project-tracking/projects/${trackingId}`, data)
  },

  // 更新任务跟踪状态
  updateTaskTracking(taskTrackingId: number, data: {
    status?: string
    current_progress?: number
    delay_days?: number
    risk_level?: string
    notes?: string
  }) {
    return axios.put(`v1/project-tracking/tasks/${taskTrackingId}`, data)
  },

  // 手动刷新统计数据
  refreshTrackingStats(trackingId?: number) {
    if (trackingId) {
      return axios.post(`v1/project-tracking/projects/${trackingId}/refresh-stats`)
    } else {
      return axios.post('v1/project-tracking/refresh-all-stats')
    }
  },

  // 获取项目跟踪统计概览
  getTrackingOverview() {
    return axios.get('v1/project-tracking/overview')
  },

  // 获取延迟项目列表
  getDelayedProjects(params?: {
    delay_threshold?: number
    page?: number
    limit?: number
  }) {
    return axios.get('v1/project-tracking/delayed-projects', { params })
  },

  // 获取高风险项目列表
  getHighRiskProjects(params?: {
    risk_level?: '高' | '中'
    page?: number
    limit?: number
  }) {
    return axios.get('v1/project-tracking/high-risk-projects', { params })
  },

  // 获取项目任务跟踪列表
  getTrackingTasks(trackingId: number) {
    return axios.get(`v1/project-tracking/project-trackings/${trackingId}/tasks`)
  },

  // 获取任务关联的日报
  getTaskRelatedReports(taskId: string) {
    return axios.get(`v1/project-tracking/task-trackings/${taskId}/reports`)
  },

  // 获取CDE评价提醒
  getCDEAlerts(trackingId: number, unreadOnly?: boolean) {
    return axios.get(`v1/project-tracking/project-trackings/${trackingId}/cde-alerts`, {
      params: { unread_only: unreadOnly }
    })
  },

  // 标记通知为已读
  markNotificationRead(notificationId: number) {
    return axios.put(`v1/project-tracking/notifications/${notificationId}/read`)
  }
}

// 类型定义
export interface ProjectTracking {
  id: number
  project_id: string
  project_name: string
  overall_progress: number
  tracking_status: string
  total_tasks: number
  completed_tasks: number
  total_reports: number
  cde_evaluations: number
  risk_level: string
  priority_level: string
  last_update_time: string
  create_time: string
  update_time: string
}

export interface TaskTracking {
  id: number
  tracking_id: number
  task_id: string
  task_name: string
  assignee: string
  assignee_id?: string
  planned_start?: string
  planned_end?: string
  actual_start?: string
  actual_end?: string
  current_progress: number
  delay_days: number
  related_reports_count: number
  cde_evaluation_count: number
  status: string
  priority_level: string
  risk_level: string
  create_time: string
  update_time: string
}

export interface TaskReportLink {
  id: number
  task_tracking_id: number
  report_id: number
  work_item_id: number
  employee_id: string
  employee_name: string
  report_date: string
  work_content: string
  hours_spent: number
  progress_contribution: number
  has_cde_evaluation: boolean
  cde_evaluation_score?: number
  cde_evaluation_content?: string
  create_time: string
  update_time: string
}

export interface TrackingNotification {
  id: number
  tracking_id: number
  notification_type: string
  priority_level: string
  title: string
  content: string
  related_task_id?: string
  related_report_id?: number
  recipient_id: string
  recipient_name: string
  recipient_role?: string
  is_read: boolean
  is_sent: boolean
  sent_time?: string
  push_methods?: string
  create_time: string
}
